<a href="https://colab.research.google.com/github/bernlkb/STQD6014-DataScience/blob/main/Week12_DataAggregation_20240107_WC.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Week 12: Data Aggregation and Group Operation**
Please refer to the textbook: **"Python for Data Analysis"** by Wes McKinney for details of this topic on **Chapter 10**

# **Start by importing NumPy and pandas**

In [61]:
import numpy as np
import pandas as pd

# **10.1 Group Operations**
## The popular ***split-apply-combine***

In [62]:
# Example DataFrame 1
rng = np.random.default_rng(seed=12345)
df = pd.DataFrame({"key1" : ["a", "a", None, "b", "b", "a", None],
                   "key2" : pd.Series([1, 2, 1, 2, 1, None, 1], dtype="Int64"),
                   "data1" : rng.standard_normal(7),
                   "data2" : rng.standard_normal(7)})
df

key1  key2     data1     data2
0     a     1 -1.423825  0.648893
1     a     2  1.263728  0.361058
2  None     1 -0.870662 -1.952863
3     b     2 -0.259173  2.347410
4     b     1 -0.075343  0.968497
5     a  <NA> -0.740885 -0.759387
6  None     1 -1.367793  0.902198

In [63]:
# Extract data1 and call groupby with the key1 column
# Create an intermediate file -> grouped
grouped = df['data1'].groupby(df['key1'])
grouped.head()

0   -1.423825
1    1.263728
3   -0.259173
4   -0.075343
5   -0.740885
Name: data1, dtype: float64

In [64]:
# Compute median for the intermediate grouped variable
# median() -> aggregation function -> 'transform' the data into single value
grouped.median()

key1
a   -0.740885
b   -0.167258
Name: data1, dtype: float64

In [65]:
# Another way
# Compute median without intermediate file
df['data1'].groupby(df['key1']).median()

key1
a   -0.740885
b   -0.167258
Name: data1, dtype: float64

In [66]:
# Another example: passing multiple groupby arrays of keys to find mean
# Passing multiple groupby arguments as a list
means = df['data1'].groupby([df['key1'], df['key2']]).mean()
means

key1  key2
a     1      -1.423825
      2       1.263728
b     1      -0.075343
      2      -0.259173
Name: data1, dtype: float64

In [67]:
# Unstack
# transform the table -> from long (vertical) to wide (horizontal) format
means.unstack()

key2         1         2
key1                    
a    -1.423825  1.263728
b    -0.075343 -0.259173

In [68]:
# Display df content
df

key1  key2     data1     data2
0     a     1 -1.423825  0.648893
1     a     2  1.263728  0.361058
2  None     1 -0.870662 -1.952863
3     b     2 -0.259173  2.347410
4     b     1 -0.075343  0.968497
5     a  <NA> -0.740885 -0.759387
6  None     1 -1.367793  0.902198

In [69]:
# Pass column name as the groupby key
# Find mean for the whole data frame
df.groupby('key1').mean()

key2     data1     data2
key1                          
a      1.5 -0.300327  0.083521
b      1.5 -0.167258  1.657953

In [70]:
# Next example: "key1" is not numeric
# The resulting DataFrame will only include the mean values data1 and data2 that are numeric
# "key1" doesn't appear in the output.
df.groupby('key2').mean(numeric_only = True)

data1     data2
key2                    
1    -0.934406  0.141681
2     0.502278  1.354234

### **Iterating over Groups**
***groupby*** supports ***iteration***, generating ***a sequence of 2-tuples***

In [71]:
# Display df content
df

key1  key2     data1     data2
0     a     1 -1.423825  0.648893
1     a     2  1.263728  0.361058
2  None     1 -0.870662 -1.952863
3     b     2 -0.259173  2.347410
4     b     1 -0.075343  0.968497
5     a  <NA> -0.740885 -0.759387
6  None     1 -1.367793  0.902198

In [72]:
# Iteration using a for loop
# To access individual data for each key1 value -> 'a' & 'b'
# name -> individual key1 values
# group -> data associated with the 'name'
for name, group in df.groupby('key1'):
    print(name)
    print(group)
    print('\n ------ \n')

a
  key1  key2     data1     data2
0    a     1 -1.423825  0.648893
1    a     2  1.263728  0.361058
5    a  <NA> -0.740885 -0.759387

 ------ 

b
  key1  key2     data1     data2
3    b     2 -0.259173  2.347410
4    b     1 -0.075343  0.968497

 ------ 



In [73]:
# Display df content
df

key1  key2     data1     data2
0     a     1 -1.423825  0.648893
1     a     2  1.263728  0.361058
2  None     1 -0.870662 -1.952863
3     b     2 -0.259173  2.347410
4     b     1 -0.075343  0.968497
5     a  <NA> -0.740885 -0.759387
6  None     1 -1.367793  0.902198

# Using tuple

In [74]:
# In the case of multiple keys
# first element will be a tuple of key values
for (k1, k2), group in df.groupby(['key1', 'key2']):
    print((k1, k2))
    print(group)
    print('\n ------ \n')

('a', 1)
  key1  key2     data1     data2
0    a     1 -1.423825  0.648893

 ------ 

('a', 2)
  key1  key2     data1     data2
1    a     2  1.263728  0.361058

 ------ 

('b', 1)
  key1  key2     data1     data2
4    b     1 -0.075343  0.968497

 ------ 

('b', 2)
  key1  key2     data1    data2
3    b     2 -0.259173  2.34741

 ------ 



# Using dictionary method -> {key:value} pair

In [75]:
# Output as a dictionary
# Compute a dictionary using one-liner
pieces = {name: group for name, group in df.groupby('key1')}
pieces

{'a':   key1  key2     data1     data2
 0    a     1 -1.423825  0.648893
 1    a     2  1.263728  0.361058
 5    a  <NA> -0.740885 -0.759387,
 'b':   key1  key2     data1     data2
 3    b     2 -0.259173  2.347410
 4    b     1 -0.075343  0.968497}

In [76]:
# Accessing individual item in dictionary
# Using the "key"
pieces['b']

key1  key2     data1     data2
3    b     2 -0.259173  2.347410
4    b     1 -0.075343  0.968497

### **Grouping with Dictionaries and Series**

In [77]:
# Example DataFrame
# setting the seed number to ensure reproducibility
rng = np.random.default_rng(seed=12345)
people = pd.DataFrame(rng.standard_normal((5, 5)),
                      columns=["a", "b", "c", "d", "e"],
                      index=["Joe", "Steve", "Wanda", "Jill", "Trey"])
people

a         b         c         d         e
Joe   -1.423825  1.263728 -0.870662 -0.259173 -0.075343
Steve -0.740885 -1.367793  0.648893  0.361058 -1.952863
Wanda  2.347410  0.968497 -0.759387  0.902198 -0.466953
Jill  -0.060690  0.788844 -1.256668  0.575858  1.398979
Trey   1.322298 -0.299699  0.902919 -1.621583 -0.158189

In [78]:
# Add a few NA values
# Using the explicit indexer -> loc method
people.loc['Wanda', ['b', 'c']] = np.nan
people

a         b         c         d         e
Joe   -1.423825  1.263728 -0.870662 -0.259173 -0.075343
Steve -0.740885 -1.367793  0.648893  0.361058 -1.952863
Wanda  2.347410       NaN       NaN  0.902198 -0.466953
Jill  -0.060690  0.788844 -1.256668  0.575858  1.398979
Trey   1.322298 -0.299699  0.902919 -1.621583 -0.158189

In [79]:
# Add a few more NA values
# Using the implicit indexer -> iloc method
# (4, 1) -> rows
# (0, 3) -> columns
people.iloc[(4, 1), (0, 3)] = np.nan

In [80]:
# Group correspondence for the same columns
# Mapping using a dictionary
mapping = {"a": "red", "b": "red", "c": "blue",
           "d": "blue", "e": "red", "f" : "orange", "g" : "violet"}

In [81]:
# Mapping and compute sum
# Using the people dataset
# Example from the people dataset: "-0.870662 - 0.259173" = -1.129835 (blue)
people.groupby(mapping, axis = 1).sum()

blue       red
Joe   -1.129835 -0.235440
Steve  0.648893 -4.061540
Wanda  0.902198  1.880456
Jill  -0.680811  2.127134
Trey  -0.718663 -0.457888

In [82]:
# Challenge: add the column f and g
# Using the assign() function
rng = np.random.default_rng(seed=12345)
people = people.assign(f = rng.standard_normal((5, 1)),
                       g = rng.standard_normal((5, 1)))
people

a         b         c         d         e         f         g
Joe   -1.423825  1.263728 -0.870662 -0.259173 -0.075343 -1.423825 -0.740885
Steve -0.740885 -1.367793  0.648893       NaN -1.952863  1.263728 -1.367793
Wanda  2.347410       NaN       NaN  0.902198 -0.466953 -0.870662  0.648893
Jill  -0.060690  0.788844 -1.256668  0.575858  1.398979 -0.259173  0.361058
Trey        NaN -0.299699  0.902919 -1.621583 -0.158189 -0.075343 -1.952863

In [83]:
# Mapping and compute sum again
people.groupby(mapping, axis = 1).sum()

blue    orange       red    violet
Joe   -1.129835 -1.423825 -0.235440 -0.740885
Steve  0.648893  1.263728 -4.061540 -1.367793
Wanda  0.902198 -0.870662  1.880456  0.648893
Jill  -0.680811 -0.259173  2.127134  0.361058
Trey  -0.718663 -0.075343 -0.457888 -1.952863

# **10.2 Data Aggregation**

* we wanted to transform the data into <font color=red>***one single value***</font>
* sum, median, etc...


In [84]:
# Display df content from previous example
df

key1  key2     data1     data2
0     a     1 -1.423825  0.648893
1     a     2  1.263728  0.361058
2  None     1 -0.870662 -1.952863
3     b     2 -0.259173  2.347410
4     b     1 -0.075343  0.968497
5     a  <NA> -0.740885 -0.759387
6  None     1 -1.367793  0.902198

In [85]:
# Create an intermediate grouped object
grouped = df.groupby('key1')

In [86]:
# Select smallest values
grouped['data1'].nsmallest(2)

key1   
a     0   -1.423825
      5   -0.740885
b     3   -0.259173
      4   -0.075343
Name: data1, dtype: float64

In [87]:
# Define a peak_to_peak functions
# getting the range -> max - min
# return -> return the computed value
def peak_to_peak(arr):
    return arr.max() - arr.min()

# ***The agg() function in pandas***

* helps us <font color='red'>***break down our data into groups***</font>
* <font color='red'>***performs calculations or apply functions to each group***</font>
* <font color='red'>***returned aggregated results -> one single value***</font>
* a powerful tool for data analysis and summarization

In [88]:
# apply the peak_to_peak function to the dataframe
# to get range of data
grouped.agg(peak_to_peak)

key2     data1     data2
key1                          
a        1  2.687553  1.408280
b        1  0.183830  1.378913

In [89]:
# Manual calculation to double check
# range of data2 for key1 = a
0.648893 - (-0.759387)

1.40828

In [90]:
# The describe() -> generate a descriptive statistical summary of each group in a grouped DataFrame
# giving a quick overview of the data distribution within each group
grouped.describe()

key2                                           data1            ...  \
     count mean       std  min   25%  50%   75%  max count      mean  ...   
key1                                                                  ...   
a      2.0  1.5  0.707107  1.0  1.25  1.5  1.75  2.0   3.0 -0.300327  ...   
b      2.0  1.5  0.707107  1.0  1.25  1.5  1.75  2.0   2.0 -0.167258  ...   

                         data2                                          \
           75%       max count      mean       std       min       25%   
key1                                                                     
a     0.261422  1.263728   3.0  0.083521  0.744032 -0.759387 -0.199165   
b    -0.121301 -0.075343   2.0  1.657953  0.975039  0.968497  1.313225   

                                    
           50%       75%       max  
key1                                
a     0.361058  0.504975  0.648893  
b     1.657953  2.002681  2.347410  

[2 rows x 24 columns]

In [91]:
# To know more information about the describe() function
grouped.describe?

## **Column-Wise and Multiple Function Application**

In [92]:
# Get the tipping dataset from github
# Can download here: https://bit.ly/3VyE0vP
tips = pd.read_csv('https://bit.ly/3VyE0vP')
tips

total_bill   tip smoker   day    time  size
0         16.99  1.01     No   Sun  Dinner     2
1         10.34  1.66     No   Sun  Dinner     3
2         21.01  3.50     No   Sun  Dinner     3
3         23.68  3.31     No   Sun  Dinner     2
4         24.59  3.61     No   Sun  Dinner     4
..          ...   ...    ...   ...     ...   ...
239       29.03  5.92     No   Sat  Dinner     3
240       27.18  2.00    Yes   Sat  Dinner     2
241       22.67  2.00    Yes   Sat  Dinner     2
242       17.82  1.75     No   Sat  Dinner     2
243       18.78  3.00     No  Thur  Dinner     2

[244 rows x 6 columns]

In [93]:
# Add additional column: tip_pct
# can use assign() also, but not necessary
tips['tip_pct'] = tips['tip'] / tips['total_bill']
tips['tip_pct_times100'] = tips['tip_pct']*100
tips

total_bill   tip smoker   day    time  size   tip_pct  tip_pct_times100
0         16.99  1.01     No   Sun  Dinner     2  0.059447          5.944673
1         10.34  1.66     No   Sun  Dinner     3  0.160542         16.054159
2         21.01  3.50     No   Sun  Dinner     3  0.166587         16.658734
3         23.68  3.31     No   Sun  Dinner     2  0.139780         13.978041
4         24.59  3.61     No   Sun  Dinner     4  0.146808         14.680765
..          ...   ...    ...   ...     ...   ...       ...               ...
239       29.03  5.92     No   Sat  Dinner     3  0.203927         20.392697
240       27.18  2.00    Yes   Sat  Dinner     2  0.073584          7.358352
241       22.67  2.00    Yes   Sat  Dinner     2  0.088222          8.822232
242       17.82  1.75     No   Sat  Dinner     2  0.098204          9.820426
243       18.78  3.00     No  Thur  Dinner     2  0.159744         15.974441

[244 rows x 8 columns]

In [94]:
# Group the tips by day and smoker
# Create an intermediate grouped object
grouped = tips.groupby(['day', 'smoker'])

In [95]:
# Extract the tip_pct from grouped object
# groupby in groupby
# Compute mean on 'tip_pct'
# Using the agg() function
# 'mean' -> is not a string
# 'mean' -> built-in function in python
# built-in -> highly optimised and have higher performance
# compared to user-defined function
grouped_pct = grouped['tip_pct']
grouped_pct.agg(['mean'])

mean
day  smoker          
Fri  No      0.151650
     Yes     0.174783
Sat  No      0.158048
     Yes     0.147906
Sun  No      0.160113
     Yes     0.187250
Thur No      0.160298
     Yes     0.163863

In [96]:
# Passing multiple aggregation functions as a list
grouped_pct.agg(['mean', 'std', peak_to_peak])

mean       std  peak_to_peak
day  smoker                                  
Fri  No      0.151650  0.028123      0.067349
     Yes     0.174783  0.051293      0.159925
Sat  No      0.158048  0.039767      0.235193
     Yes     0.147906  0.061375      0.290095
Sun  No      0.160113  0.042347      0.193226
     Yes     0.187250  0.154134      0.644685
Thur No      0.160298  0.038774      0.193350
     Yes     0.163863  0.039389      0.151240

In [97]:
# Give customised name to the list of aggregated function
grouped_pct.agg([('Purata', np.mean), ('STDevvvv', np.std)])

Purata  STDevvvv
day  smoker                    
Fri  No      0.151650  0.028123
     Yes     0.174783  0.051293
Sat  No      0.158048  0.039767
     Yes     0.147906  0.061375
Sun  No      0.160113  0.042347
     Yes     0.187250  0.154134
Thur No      0.160298  0.038774
     Yes     0.163863  0.039389

In [98]:
# Specify a list of functions to apply
# 'tip_pct' and 'total_bill'
funcs = ['count', 'mean', 'max', 'min', 'median']
res = grouped[['tip_pct', 'total_bill']].agg(funcs)
res

tip_pct                                         total_bill  \
              count      mean       max       min    median      count   
day  smoker                                                              
Fri  No           4  0.151650  0.187735  0.120385  0.149241          4   
     Yes         15  0.174783  0.263480  0.103555  0.173913         15   
Sat  No          45  0.158048  0.291990  0.056797  0.150152         45   
     Yes         42  0.147906  0.325733  0.035638  0.153624         42   
Sun  No          57  0.160113  0.252672  0.059447  0.161665         57   
     Yes         19  0.187250  0.710345  0.065660  0.138122         19   
Thur No          45  0.160298  0.266312  0.072961  0.153492         45   
     Yes         17  0.163863  0.241255  0.090014  0.153846         17   

                                              
                  mean    max    min  median  
day  smoker                                   
Fri  No      18.420000  22.75  12.46  19.235  
     Yes     16.813333  40.17   5.75  13.420  
Sat  No      19.661778  48.33   7.25  17.820  
     Yes     21.276667  50.81   3.07  20.390  
Sun  No      20.506667  48.17   8.77  18.430  
     Yes     24.120000  45.35   7.25  23.100  
Thur No      17.113111  41.19   7.51  15.950  
     Yes     19.190588  43.11  10.34  16.470

In [99]:
# Accessing individual data from aggregation result
# 'total_bill'
res['total_bill']

count       mean    max    min  median
day  smoker                                        
Fri  No          4  18.420000  22.75  12.46  19.235
     Yes        15  16.813333  40.17   5.75  13.420
Sat  No         45  19.661778  48.33   7.25  17.820
     Yes        42  21.276667  50.81   3.07  20.390
Sun  No         57  20.506667  48.17   8.77  18.430
     Yes        19  24.120000  45.35   7.25  23.100
Thur No         45  17.113111  41.19   7.51  15.950
     Yes        17  19.190588  43.11  10.34  16.470

In [100]:
# Another way: tuples
# Can pass aggreation functions as a list of tuples
tuples = [('AvErages', 'mean'), ('VaRianCE', np.var)]
%timeit grouped[['tip_pct', 'total_bill']].agg(tuples)

4.32 ms ± 594 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [101]:
# Another way: tuples
# Can pass aggreation functions as a list of tuples
tuples = [('AvErages', np.mean), ('VaRianCE', np.var)]
%timeit grouped[['tip_pct', 'total_bill']].agg(tuples)

5.03 ms ± 173 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [102]:
# Another way: dictionary
# Applying different functions using a dictionary
grouped.agg({'tip': np.max,
             'size': 'sum'})

tip  size
day  smoker             
Fri  No       3.50     9
     Yes      4.73    31
Sat  No       9.00   115
     Yes     10.00   104
Sun  No       6.00   167
     Yes      6.50    49
Thur No       6.70   112
     Yes      5.00    40

In [103]:
# Passing aggreation functions as dictionary -> as key-values pair
grouped.agg({'total_bill': ['min', 'max', 'mean', 'std'],
             'size': 'sum'})

total_bill                              size
                   min    max       mean        std  sum
day  smoker                                             
Fri  No          12.46  22.75  18.420000   5.059282    9
     Yes          5.75  40.17  16.813333   9.086388   31
Sat  No           7.25  48.33  19.661778   8.939181  115
     Yes          3.07  50.81  21.276667  10.069138  104
Sun  No           8.77  48.17  20.506667   8.130189  167
     Yes          7.25  45.35  24.120000  10.442511   49
Thur No           7.51  41.19  17.113111   7.721728  112
     Yes         10.34  43.11  19.190588   8.355149   40

## **Returning Aggregated Data Without Row Indexes**

In [104]:
# Display tips dataframe
tips

total_bill   tip smoker   day    time  size   tip_pct  tip_pct_times100
0         16.99  1.01     No   Sun  Dinner     2  0.059447          5.944673
1         10.34  1.66     No   Sun  Dinner     3  0.160542         16.054159
2         21.01  3.50     No   Sun  Dinner     3  0.166587         16.658734
3         23.68  3.31     No   Sun  Dinner     2  0.139780         13.978041
4         24.59  3.61     No   Sun  Dinner     4  0.146808         14.680765
..          ...   ...    ...   ...     ...   ...       ...               ...
239       29.03  5.92     No   Sat  Dinner     3  0.203927         20.392697
240       27.18  2.00    Yes   Sat  Dinner     2  0.073584          7.358352
241       22.67  2.00    Yes   Sat  Dinner     2  0.088222          8.822232
242       17.82  1.75     No   Sat  Dinner     2  0.098204          9.820426
243       18.78  3.00     No  Thur  Dinner     2  0.159744         15.974441

[244 rows x 8 columns]

In [105]:
# Need to pass "numeric_only = True" to mean() as it involves non-numeric data
tips.groupby(['day', 'smoker']).mean(numeric_only = True)

total_bill       tip      size   tip_pct  tip_pct_times100
day  smoker                                                            
Fri  No       18.420000  2.812500  2.250000  0.151650         15.165044
     Yes      16.813333  2.714000  2.066667  0.174783         17.478305
Sat  No       19.661778  3.102889  2.555556  0.158048         15.804766
     Yes      21.276667  2.875476  2.476190  0.147906         14.790607
Sun  No       20.506667  3.167895  2.929825  0.160113         16.011294
     Yes      24.120000  3.516842  2.578947  0.187250         18.725032
Thur No       17.113111  2.673778  2.488889  0.160298         16.029808
     Yes      19.190588  3.030000  2.352941  0.163863         16.386327

In [106]:
# If we want to disable the index
tips.groupby(['day', 'smoker'], as_index=False).mean(numeric_only = True)

day smoker  total_bill       tip      size   tip_pct  tip_pct_times100
0   Fri     No   18.420000  2.812500  2.250000  0.151650         15.165044
1   Fri    Yes   16.813333  2.714000  2.066667  0.174783         17.478305
2   Sat     No   19.661778  3.102889  2.555556  0.158048         15.804766
3   Sat    Yes   21.276667  2.875476  2.476190  0.147906         14.790607
4   Sun     No   20.506667  3.167895  2.929825  0.160113         16.011294
5   Sun    Yes   24.120000  3.516842  2.578947  0.187250         18.725032
6  Thur     No   17.113111  2.673778  2.488889  0.160298         16.029808
7  Thur    Yes   19.190588  3.030000  2.352941  0.163863         16.386327

# **10.3 Apply: General split-apply-combine**

In [107]:
# Define a function called 'top'
# select the top five tip_pct values by group
def top(df, n = 5, column = 'tip_pct'):
    return df.sort_values(column, ascending = False)[:n]

In [108]:
# Display tips content
tips

total_bill   tip smoker   day    time  size   tip_pct  tip_pct_times100
0         16.99  1.01     No   Sun  Dinner     2  0.059447          5.944673
1         10.34  1.66     No   Sun  Dinner     3  0.160542         16.054159
2         21.01  3.50     No   Sun  Dinner     3  0.166587         16.658734
3         23.68  3.31     No   Sun  Dinner     2  0.139780         13.978041
4         24.59  3.61     No   Sun  Dinner     4  0.146808         14.680765
..          ...   ...    ...   ...     ...   ...       ...               ...
239       29.03  5.92     No   Sat  Dinner     3  0.203927         20.392697
240       27.18  2.00    Yes   Sat  Dinner     2  0.073584          7.358352
241       22.67  2.00    Yes   Sat  Dinner     2  0.088222          8.822232
242       17.82  1.75     No   Sat  Dinner     2  0.098204          9.820426
243       18.78  3.00     No  Thur  Dinner     2  0.159744         15.974441

[244 rows x 8 columns]

In [109]:
# Apply the top function
# Return top six
# new argument will overwrite the default value in the function
# 'tip_pct' is the default column
top(tips, n = 6)

total_bill   tip smoker  day    time  size   tip_pct  tip_pct_times100
172        7.25  5.15    Yes  Sun  Dinner     2  0.710345         71.034483
178        9.60  4.00    Yes  Sun  Dinner     2  0.416667         41.666667
67         3.07  1.00    Yes  Sat  Dinner     1  0.325733         32.573290
232       11.61  3.39     No  Sat  Dinner     2  0.291990         29.198966
183       23.17  6.50    Yes  Sun  Dinner     4  0.280535         28.053517
109       14.31  4.00    Yes  Sat  Dinner     2  0.279525         27.952481

In [110]:
# groupby 'smoker'
# return top 5 for each 'smoker' group -> No and Yes
tips.groupby('smoker').apply(top)

total_bill   tip smoker   day    time  size   tip_pct  \
smoker                                                              
No     232       11.61  3.39     No   Sat  Dinner     2  0.291990   
       149        7.51  2.00     No  Thur   Lunch     2  0.266312   
       51        10.29  2.60     No   Sun  Dinner     2  0.252672   
       185       20.69  5.00     No   Sun  Dinner     5  0.241663   
       88        24.71  5.85     No  Thur   Lunch     2  0.236746   
Yes    172        7.25  5.15    Yes   Sun  Dinner     2  0.710345   
       178        9.60  4.00    Yes   Sun  Dinner     2  0.416667   
       67         3.07  1.00    Yes   Sat  Dinner     1  0.325733   
       183       23.17  6.50    Yes   Sun  Dinner     4  0.280535   
       109       14.31  4.00    Yes   Sat  Dinner     2  0.279525   

            tip_pct_times100  
smoker                        
No     232         29.198966  
       149         26.631158  
       51          25.267250  
       185         24.166264  
       88          23.674626  
Yes    172         71.034483  
       178         41.666667  
       67          32.573290  
       183         28.053517  
       109         27.952481

In [111]:
# Pass additional argument after the function
# Overwrite the default values
tips.groupby(['smoker', 'day']).apply(top, n = 3, column = 'total_bill')

total_bill    tip smoker   day    time  size   tip_pct  \
smoker day                                                                
No     Fri  94        22.75   3.25     No   Fri  Dinner     2  0.142857   
            91        22.49   3.50     No   Fri  Dinner     2  0.155625   
            223       15.98   3.00     No   Fri   Lunch     3  0.187735   
       Sat  212       48.33   9.00     No   Sat  Dinner     4  0.186220   
            59        48.27   6.73     No   Sat  Dinner     4  0.139424   
            23        39.42   7.58     No   Sat  Dinner     4  0.192288   
       Sun  156       48.17   5.00     No   Sun  Dinner     6  0.103799   
            112       38.07   4.00     No   Sun  Dinner     3  0.105070   
            11        35.26   5.00     No   Sun  Dinner     4  0.141804   
       Thur 142       41.19   5.00     No  Thur   Lunch     5  0.121389   
            85        34.83   5.17     No  Thur   Lunch     4  0.148435   
            141       34.30   6.70     No  Thur   Lunch     6  0.195335   
Yes    Fri  95        40.17   4.73    Yes   Fri  Dinner     4  0.117750   
            90        28.97   3.00    Yes   Fri  Dinner     2  0.103555   
            96        27.28   4.00    Yes   Fri  Dinner     2  0.146628   
       Sat  170       50.81  10.00    Yes   Sat  Dinner     3  0.196812   
            102       44.30   2.50    Yes   Sat  Dinner     3  0.056433   
            207       38.73   3.00    Yes   Sat  Dinner     4  0.077459   
       Sun  182       45.35   3.50    Yes   Sun  Dinner     3  0.077178   
            184       40.55   3.00    Yes   Sun  Dinner     2  0.073983   
            180       34.65   3.68    Yes   Sun  Dinner     4  0.106205   
       Thur 197       43.11   5.00    Yes  Thur   Lunch     4  0.115982   
            83        32.68   5.00    Yes  Thur   Lunch     2  0.152999   
            192       28.44   2.56    Yes  Thur   Lunch     2  0.090014   

                 tip_pct_times100  
smoker day                         
No     Fri  94          14.285714  
            91          15.562472  
            223         18.773467  
       Sat  212         18.621974  
            59          13.942407  
            23          19.228818  
       Sun  156         10.379905  
            112         10.506961  
            11          14.180374  
       Thur 142         12.138869  
            85          14.843526  
            141         19.533528  
Yes    Fri  95          11.774956  
            90          10.355540  
            96          14.662757  
       Sat  170         19.681165  
            102          5.643341  
            207          7.745933  
       Sun  182          7.717751  
            184          7.398274  
            180         10.620491  
       Thur 197         11.598237  
            83          15.299878  
            192          9.001406

In [112]:
# Using describe() function to get descriptive statistics
# In wide table format
res = tips.groupby('smoker')['tip_pct'].describe()
res

count      mean       std       min       25%       50%       75%  \
smoker                                                                      
No      151.0  0.159328  0.039910  0.056797  0.136906  0.155625  0.185014   
Yes      93.0  0.163196  0.085119  0.035638  0.106771  0.153846  0.195059   

             max  
smoker            
No      0.291990  
Yes     0.710345

In [113]:
# Unstack
# Transform wide table to long format
res.unstack('smoker')

       smoker
count  No        151.000000
       Yes        93.000000
mean   No          0.159328
       Yes         0.163196
std    No          0.039910
       Yes         0.085119
min    No          0.056797
       Yes         0.035638
25%    No          0.136906
       Yes         0.106771
50%    No          0.155625
       Yes         0.153846
75%    No          0.185014
       Yes         0.195059
max    No          0.291990
       Yes         0.710345
dtype: float64

## **Suppressing the Group Keys**

In [114]:
# Suppressing group key -> group_keys function
tips.groupby('smoker', group_keys=False).apply(top)

total_bill   tip smoker   day    time  size   tip_pct  tip_pct_times100
232       11.61  3.39     No   Sat  Dinner     2  0.291990         29.198966
149        7.51  2.00     No  Thur   Lunch     2  0.266312         26.631158
51        10.29  2.60     No   Sun  Dinner     2  0.252672         25.267250
185       20.69  5.00     No   Sun  Dinner     5  0.241663         24.166264
88        24.71  5.85     No  Thur   Lunch     2  0.236746         23.674626
172        7.25  5.15    Yes   Sun  Dinner     2  0.710345         71.034483
178        9.60  4.00    Yes   Sun  Dinner     2  0.416667         41.666667
67         3.07  1.00    Yes   Sat  Dinner     1  0.325733         32.573290
183       23.17  6.50    Yes   Sun  Dinner     4  0.280535         28.053517
109       14.31  4.00    Yes   Sat  Dinner     2  0.279525         27.952481

## **Quantile and Bucket Analysis**

In [115]:
# Example dataset
rng = np.random.default_rng(seed=12345)
frame = pd.DataFrame({"data1": rng.standard_normal(1000),
                      "data2": rng.standard_normal(1000)})
frame.head()

data1     data2
0 -1.423825 -1.176622
1  1.263728  1.205009
2 -0.870662  0.717879
3 -0.259173 -2.686273
4 -0.075343 -1.865260

In [116]:
# Using the cut() function to transform frame into respective bin -> categorical data
# '( ]' -> left bound is exclusive, right bound is inclusive
quartiles = pd.cut(frame['data1'], 4)
quartiles.head()

0    (-1.537, 0.0459]
1     (0.0459, 1.629]
2    (-1.537, 0.0459]
3    (-1.537, 0.0459]
4    (-1.537, 0.0459]
Name: data1, dtype: category
Categories (4, interval[float64, right]): [(-3.126, -1.537] < (-1.537, 0.0459] < (0.0459, 1.629] <
                                           (1.629, 3.211]]

In [117]:
# Create a new column in the DataFrame to store quartile information
frame['quartiles'] = quartiles
frame.head()

data1     data2         quartiles
0 -1.423825 -1.176622  (-1.537, 0.0459]
1  1.263728  1.205009   (0.0459, 1.629]
2 -0.870662  0.717879  (-1.537, 0.0459]
3 -0.259173 -2.686273  (-1.537, 0.0459]
4 -0.075343 -1.865260  (-1.537, 0.0459]

In [118]:
# Group the DataFrame by quartiles and access 'data1' values in each group
grouped_data = frame.groupby('quartiles')['data1']

# Filter the DataFrame to get 'data1' values for the first quartile
# explicit index
first_quartile_data1 = frame.loc[frame['quartiles'] == quartiles.cat.categories[0], 'data1']
first_quartile_data1.head()

9    -1.952863
23   -1.621583
57   -1.602702
64   -1.640418
82   -1.795713
Name: data1, dtype: float64

In [119]:
# Min for data1 in first quartile bin label
print('Min: ' + str(first_quartile_data1.min()))

# Max for data1 in first quartile bin label
print('Max: ' + str(first_quartile_data1.max()))

# Count for data1 in first quartile bin label
print('Count: ' + str(first_quartile_data1.count()))

# Mean for data1 in first quartile bin label
print('Mean: ' + str(first_quartile_data1.mean()))

Min: -3.1196086260734055
Max: -1.538154034502331
Count: 65
Mean: -2.0196746416859885


In [120]:
# Define a function -> get_stats
def get_stats(group):
    return pd.DataFrame(
        {'min': group.min(), 'max': group.max(),
         'count': group.count(), 'mean': group.mean()}
    )

In [121]:
# Intermediate object
# quartiles -> four categorical bins
grouped = frame.groupby(quartiles)

In [122]:
# Apply the get_stats function
grouped.apply(get_stats)

<ipython-input-120-322b6e1a6b5a>:5: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  'count': group.count(), 'mean': group.mean()}
<ipython-input-120-322b6e1a6b5a>:5: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  'count': group.count(), 'mean': group.mean()}
<ipython-input-120-322b6e1a6b5a>:5: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeri

min               max  count  \
data1                                                                   
(-3.126, -1.537] data1             -3.119609         -1.538154     65   
                 data2             -2.433466          2.040796     65   
                 quartiles  (-3.126, -1.537]  (-3.126, -1.537]     65   
(-1.537, 0.0459] data1             -1.523449          0.045684    443   
                 data2             -2.830704          3.721098    443   
                 quartiles  (-1.537, 0.0459]  (-1.537, 0.0459]    443   
(0.0459, 1.629]  data1              0.048765          1.621952    438   
                 data2             -2.320815          3.461976    438   
                 quartiles   (0.0459, 1.629]   (0.0459, 1.629]    438   
(1.629, 3.211]   data1              1.651471          3.211418     54   
                 data2             -1.778229          2.530599     54   
                 quartiles    (1.629, 3.211]    (1.629, 3.211]     54   

                                mean  
data1                                 
(-3.126, -1.537] data1     -2.019675  
                 data2      0.033381  
                 quartiles       NaN  
(-1.537, 0.0459] data1     -0.605092  
                 data2      0.020211  
                 quartiles       NaN  
(0.0459, 1.629]  data1      0.678524  
                 data2      0.029518  
                 quartiles       NaN  
(1.629, 3.211]   data1      2.044438  
                 data2      0.228316  
                 quartiles       NaN

In [123]:
# Using lists of pre-defined standard python functions
# ['min', 'max', 'count', 'mean'] -> built-in function (optimised in C language)
grouped.agg(['min', 'max', 'count', 'mean'])

<ipython-input-123-c20272b88d0d>:3: FutureWarning: ['quartiles'] did not aggregate successfully. If any error is raised this will raise in a future version of pandas. Drop these columns/ops to avoid this warning.
  grouped.agg(['min', 'max', 'count', 'mean'])


data1                               data2            \
                       min       max count      mean       min       max   
data1                                                                      
(-3.126, -1.537] -3.119609 -1.538154    65 -2.019675 -2.433466  2.040796   
(-1.537, 0.0459] -1.523449  0.045684   443 -0.605092 -2.830704  3.721098   
(0.0459, 1.629]   0.048765  1.621952   438  0.678524 -2.320815  3.461976   
(1.629, 3.211]    1.651471  3.211418    54  2.044438 -1.778229  2.530599   

                                  
                 count      mean  
data1                             
(-3.126, -1.537]    65  0.033381  
(-1.537, 0.0459]   443  0.020211  
(0.0459, 1.629]    438  0.029518  
(1.629, 3.211]      54  0.228316

In [124]:
# Display top few rows of frame
frame.head()

data1     data2         quartiles
0 -1.423825 -1.176622  (-1.537, 0.0459]
1  1.263728  1.205009   (0.0459, 1.629]
2 -0.870662  0.717879  (-1.537, 0.0459]
3 -0.259173 -2.686273  (-1.537, 0.0459]
4 -0.075343 -1.865260  (-1.537, 0.0459]

In [125]:
pd.qcut?

In [126]:
pd.cut?

In [127]:
# Assign each data into respective bin
# Dividing data1 into 4 bins based on their quantiles
quartiles_samp = pd.qcut(frame['data1'], 4)
quartiles_samp

0      (-3.121, -0.675]
1        (0.687, 3.211]
2      (-3.121, -0.675]
3      (-0.675, 0.0134]
4      (-0.675, 0.0134]
             ...       
995     (0.0134, 0.687]
996     (0.0134, 0.687]
997    (-0.675, 0.0134]
998     (0.0134, 0.687]
999    (-0.675, 0.0134]
Name: data1, Length: 1000, dtype: category
Categories (4, interval[float64, right]): [(-3.121, -0.675] < (-0.675, 0.0134] < (0.0134, 0.687] <
                                           (0.687, 3.211]]

In [128]:
i = 0
for i, bin_interval in enumerate(quartiles_samp):
    print(f'Bin {i + 1}: {bin_interval}')

Bin 1: (-3.121, -0.675]
Bin 2: (0.687, 3.211]
Bin 3: (-3.121, -0.675]
Bin 4: (-0.675, 0.0134]
Bin 5: (-0.675, 0.0134]
Bin 6: (-3.121, -0.675]
Bin 7: (-3.121, -0.675]
Bin 8: (0.0134, 0.687]
Bin 9: (0.0134, 0.687]
Bin 10: (-3.121, -0.675]
Bin 11: (0.687, 3.211]
Bin 12: (0.687, 3.211]
Bin 13: (-3.121, -0.675]
Bin 14: (0.687, 3.211]
Bin 15: (-0.675, 0.0134]
Bin 16: (-0.675, 0.0134]
Bin 17: (0.687, 3.211]
Bin 18: (-3.121, -0.675]
Bin 19: (0.0134, 0.687]
Bin 20: (0.687, 3.211]
Bin 21: (0.687, 3.211]
Bin 22: (-0.675, 0.0134]
Bin 23: (0.687, 3.211]
Bin 24: (-3.121, -0.675]
Bin 25: (-0.675, 0.0134]
Bin 26: (0.0134, 0.687]
Bin 27: (-3.121, -0.675]
Bin 28: (-0.675, 0.0134]
Bin 29: (0.687, 3.211]
Bin 30: (0.687, 3.211]
Bin 31: (0.687, 3.211]
Bin 32: (0.687, 3.211]
Bin 33: (-3.121, -0.675]
Bin 34: (-3.121, -0.675]
Bin 35: (-3.121, -0.675]
Bin 36: (0.0134, 0.687]
Bin 37: (0.687, 3.211]
Bin 38: (-0.675, 0.0134]
Bin 39: (-3.121, -0.675]
Bin 40: (0.0134, 0.687]
Bin 41: (0.0134, 0.687]
Bin 42: (-0.675, 

In [129]:
# Apply the get_stats function
# get_stats -> min, max, count & mean
grouped = frame.groupby(quartiles_samp)
grouped.apply(get_stats)

<ipython-input-120-322b6e1a6b5a>:5: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  'count': group.count(), 'mean': group.mean()}
<ipython-input-120-322b6e1a6b5a>:5: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  'count': group.count(), 'mean': group.mean()}
<ipython-input-120-322b6e1a6b5a>:5: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeri

min               max  count  \
data1                                                                   
(-3.121, -0.675] data1             -3.119609         -0.678494    250   
                 data2             -2.830704          3.041653    250   
                 quartiles  (-3.126, -1.537]  (-1.537, 0.0459]    250   
(-0.675, 0.0134] data1             -0.673305          0.008009    250   
                 data2             -2.686273          3.721098    250   
                 quartiles  (-1.537, 0.0459]  (-1.537, 0.0459]    250   
(0.0134, 0.687]  data1              0.018753          0.686183    250   
                 data2             -2.320804          3.461976    250   
                 quartiles  (-1.537, 0.0459]   (0.0459, 1.629]    250   
(0.687, 3.211]   data1              0.688282          3.211418    250   
                 data2             -2.320815          3.145451    250   
                 quartiles   (0.0459, 1.629]    (1.629, 3.211]    250   

                                mean  
data1                                 
(-3.121, -0.675] data1     -1.294347  
                 data2      0.062349  
                 quartiles       NaN  
(-0.675, 0.0134] data1     -0.303875  
                 data2     -0.016510  
                 quartiles       NaN  
(0.0134, 0.687]  data1      0.343512  
                 data2     -0.025011  
                 quartiles       NaN  
(0.687, 3.211]   data1      1.287744  
                 data2      0.124697  
                 quartiles       NaN

## **Filling Missing Values with Group-Specific Values**

* ***Data imputation*** -> fill up the missing data in the same group with certain values (can be mean, median, etc....)

In [130]:
# Example data
rng = np.random.default_rng(12345)
s = pd.Series(rng.standard_normal(6))
s

0   -1.423825
1    1.263728
2   -0.870662
3   -0.259173
4   -0.075343
5   -0.740885
dtype: float64

In [131]:
# Introduce NaN: skip of two
s[::2] = np.nan
s

0         NaN
1    1.263728
2         NaN
3   -0.259173
4         NaN
5   -0.740885
dtype: float64

In [132]:
# Compute mean of s
s.mean()

0.08789019036970125

In [133]:
# Fill NA with group mean
s.fillna(s.mean())

0    0.087890
1    1.263728
2    0.087890
3   -0.259173
4    0.087890
5   -0.740885
dtype: float64

In [134]:
# Another data
states = ["Ohio", "New York", "Vermont", "Florida",
          "Oregon", "Nevada", "California", "Idaho"]
group_key = ["East", "East", "East", "East",
             "West", "West", "West", "West"]
data = pd.Series(rng.standard_normal(8), index=states)
data

Ohio         -1.367793
New York      0.648893
Vermont       0.361058
Florida      -1.952863
Oregon        2.347410
Nevada        0.968497
California   -0.759387
Idaho         0.902198
dtype: float64

In [135]:
# Data wrangling / transformation
data[['Vermont', 'Nevada', 'Idaho']] = np.nan
data

Ohio         -1.367793
New York      0.648893
Vermont            NaN
Florida      -1.952863
Oregon        2.347410
Nevada             NaN
California   -0.759387
Idaho              NaN
dtype: float64

In [136]:
# Size based on the 'group_key'
data.groupby(group_key).size()

East    4
West    4
dtype: int64

In [137]:
# Display data content
data

Ohio         -1.367793
New York      0.648893
Vermont            NaN
Florida      -1.952863
Oregon        2.347410
Nevada             NaN
California   -0.759387
Idaho              NaN
dtype: float64

In [138]:
# count based on the 'group_key'
# Nan is not counted
data.groupby(group_key).count()

East    3
West    2
dtype: int64

In [139]:
# Define a custom function -> fill_mean
def fill_mean(group):
    return group.fillna(group.mean())

In [140]:
# Display content of data
data

Ohio         -1.367793
New York      0.648893
Vermont            NaN
Florida      -1.952863
Oregon        2.347410
Nevada             NaN
California   -0.759387
Idaho              NaN
dtype: float64

In [141]:
# Fill the NA values using the group means
data.groupby(group_key, group_keys=True).apply(fill_mean)

East  Ohio         -1.367793
      New York      0.648893
      Vermont      -0.890588
      Florida      -1.952863
West  Oregon        2.347410
      Nevada        0.794011
      California   -0.759387
      Idaho         0.794011
dtype: float64

In [142]:
# Manual calculation
(-1.367793 + 0.648893 - 1.952863) / 3

-0.8905876666666667

# **Group Weighted Average and Correlation**

In [143]:
# An example data
rng = np.random.RandomState(12345)
df = pd.DataFrame({"category": ["a", "a", "a", "a",
                                "b", "b", "b", "b"],
                   "data": rng.standard_normal(8),
                   "weights": rng.uniform(size=8)})
df

category      data   weights
0        a -0.204708  0.748907
1        a  0.478943  0.653570
2        a -0.519439  0.747715
3        a -0.555730  0.961307
4        b  1.965781  0.008388
5        b  1.393406  0.106444
6        b  0.092908  0.298704
7        b  0.281746  0.656411

In [144]:
# Intermediate object groupby category
grouped = df.groupby('category')

# Define a weighted average function -> get_wavg
def get_wavg(group):
    return np.average(group['data'], weights=group['weights'])

In [145]:
# Apply get_wavg function
grouped.apply(get_wavg)

category
a   -0.245188
b    0.352824
dtype: float64

# **Weighted Mean Formula**:

* find the ***sum of all the variables multiplied by their weight***, then ***divide by the sum of the weights***.

* For more information about Weighted mean, visit this website: http://tinyurl.com/4d4nr7fh

In [146]:
# Display content of df
df

category      data   weights
0        a -0.204708  0.748907
1        a  0.478943  0.653570
2        a -0.519439  0.747715
3        a -0.555730  0.961307
4        b  1.965781  0.008388
5        b  1.393406  0.106444
6        b  0.092908  0.298704
7        b  0.281746  0.656411

In [147]:
# Calculate weighted mean manually
(-0.204708*0.748907 + 0.478943*0.653570 + (-0.519439* 0.747715) + (-0.555730*0.961307)) / (0.748907+ 0.653570 + 0.747715 + 0.961307)

-0.24518855658992655

In [148]:
# Another example
# download the stock_px.csv here: https://bit.ly/3vuuejJ
# end-of-day prices for a Apple, Microsoft, ExxonMobil and the S&P 500 index
# parse_dates=True -> parse any columns that look like dates into datetime objects
# index_col=0 -> specifies which column from the CSV file should be used as the index
close_px = pd.read_csv('https://bit.ly/3vuuejJ',
                       parse_dates=True,
                       index_col = 0)
close_px

AAPL   MSFT    XOM      SPX
2003-01-02    7.40  21.11  29.22   909.03
2003-01-03    7.45  21.14  29.24   908.59
2003-01-06    7.45  21.52  29.96   929.01
2003-01-07    7.43  21.93  28.95   922.93
2003-01-08    7.28  21.31  28.83   909.93
...            ...    ...    ...      ...
2011-10-10  388.81  26.94  76.28  1194.89
2011-10-11  400.29  27.00  76.27  1195.54
2011-10-12  402.19  26.96  77.16  1207.25
2011-10-13  408.43  27.18  76.37  1203.66
2011-10-14  422.00  27.27  78.11  1224.58

[2214 rows x 4 columns]

In [149]:
# Check top few rows of close_px
close_px.head()

AAPL   MSFT    XOM     SPX
2003-01-02  7.40  21.11  29.22  909.03
2003-01-03  7.45  21.14  29.24  908.59
2003-01-06  7.45  21.52  29.96  929.01
2003-01-07  7.43  21.93  28.95  922.93
2003-01-08  7.28  21.31  28.83  909.93

In [150]:
# Define a custom spx_corr function
# pair-wise correlation of each column with the "SPX" column
def spx_corr(group):
    return group.corrwith(group['SPX'])

In [151]:
# pct_change() -> Calculate percentage change between the current and prior value
rets = close_px.pct_change().dropna()
rets

AAPL      MSFT       XOM       SPX
2003-01-03  0.006757  0.001421  0.000684 -0.000484
2003-01-06  0.000000  0.017975  0.024624  0.022474
2003-01-07 -0.002685  0.019052 -0.033712 -0.006545
2003-01-08 -0.020188 -0.028272 -0.004145 -0.014086
2003-01-09  0.008242  0.029094  0.021159  0.019386
...              ...       ...       ...       ...
2011-10-10  0.051406  0.026286  0.036977  0.034125
2011-10-11  0.029526  0.002227 -0.000131  0.000544
2011-10-12  0.004747 -0.001481  0.011669  0.009795
2011-10-13  0.015515  0.008160 -0.010238 -0.002974
2011-10-14  0.033225  0.003311  0.022784  0.017380

[2213 rows x 4 columns]

In [152]:
# Define a custom get_year function
def get_year(x):
    return x.year

In [153]:
# Group by year
# by_year is an intermediate object
by_year = rets.groupby(get_year)

In [154]:
# Apply spx_corr function
by_year.apply(spx_corr)

AAPL      MSFT       XOM  SPX
2003  0.541124  0.745174  0.661265  1.0
2004  0.374283  0.588531  0.557742  1.0
2005  0.467540  0.562374  0.631010  1.0
2006  0.428267  0.406126  0.518514  1.0
2007  0.508118  0.658770  0.786264  1.0
2008  0.681434  0.804626  0.828303  1.0
2009  0.707103  0.654902  0.797921  1.0
2010  0.710105  0.730118  0.839057  1.0
2011  0.691931  0.800996  0.859975  1.0

In [155]:
# Define a custom corr_aapl_msft function
# compute inter-column correlations
def corr_aapl_msft(group):
    return group['AAPL'].corr(group['MSFT'])

In [156]:
# Annual correlation between Apple and Microsoft
by_year.apply(corr_aapl_msft)

2003    0.480868
2004    0.259024
2005    0.300093
2006    0.161735
2007    0.417738
2008    0.611901
2009    0.432738
2010    0.571946
2011    0.581987
dtype: float64

# **Group-Wise Linear Regression**

In [157]:
# Importing the statsmodels library
# statsmodels -> Python library to do statistical analysis
import statsmodels.api as sm

In [158]:
# Define a custom ordinary least squares (OLS) regression function
# to find the best-fitting line by minimizing the sum of squared differences between observed and predicted values.
def regress(data, yvar=None, xvars=None):
    Y = data[yvar]
    X = data[xvars]
    X["intercept"] = 1.
    result = sm.OLS(Y, X).fit()
    return result.params

In [159]:
# Yearly linear regression of AAPL on SPX returns
# AAPL is part of Magnificent 7 in SPX500
# Apple (AAPL), Alphabet (GOOGL, GOOG), Microsoft (MSFT), Amazon (AMZN), Meta (META), Nvidia (NVDA), & Tesla (TSLA)
# AAPL = .00071 + 1.195406SPX
# Refer to this link for AAPL stock price between 2007 and 2009: http://tinyurl.com/4cs9ppp2
by_year.apply(regress, yvar = 'AAPL', xvars = ['SPX'])

SPX  intercept
2003  1.195406   0.000710
2004  1.363463   0.004201
2005  1.766415   0.003246
2006  1.645496   0.000080
2007  1.198761   0.003438
2008  0.968016  -0.001110
2009  0.879103   0.002954
2010  1.052608   0.001261
2011  0.806605   0.001514

# **10.4 Group Transforms and “Unwrapped” GroupBys**

In [160]:
# An example
df = pd.DataFrame({'key': ['a', 'b', 'c'] * 4,
                   'value': np.arange(12.)})
df

key  value
0    a    0.0
1    b    1.0
2    c    2.0
3    a    3.0
4    b    4.0
5    c    5.0
6    a    6.0
7    b    7.0
8    c    8.0
9    a    9.0
10   b   10.0
11   c   11.0

In [161]:
# group means by key
g = df.groupby('key', group_keys=True)['value'] # create intermediate object
g.mean() # compute mean

key
a    4.5
b    5.5
c    6.5
Name: value, dtype: float64

In [162]:
# Define a custom normalization function
def normalise(x):
    return (x - x.mean()) / x.std()

In [163]:
# method 1 - slower (user-defined function, preseumably not optimised)
%timeit g.apply(normalise)

2.5 ms ± 111 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [164]:
# method 2 - fast (user-defined function, preseumably not optimised)
%timeit g.transform(normalise)

2.42 ms ± 124 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [165]:
# Faster
# unwrapped group operations - doing arithmetic between the outputs of
# multiple GroupBy operations
# often much faster than a general apply function
%timeit (df['value'] - g.transform('mean')) / g.transform('std')

917 µs ± 60.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


# ***Traditional group operation***

* involve aggregating the data within each group using functions like mean, sum, etc.
* These functions reduce the data within each group to a single value (the aggregate)

# ***Unwrapped group operation***

* refer to the ability to perform operations on grouped data ***without explicitly aggregating the results***
* apply a function to each group in a DataFrame, and the function can return a ***modified version*** of the data ***rather than a single aggregated value***



# ***Still built-in aggregation functions are the fastest***
* they are ***implemented in C and optimized for performance***

In [166]:
# Fastest - using default python function
# often highly optimized and can be very fast
# built-in aggregation functions are likely to be faster because they are implemented in C and optimized for performance
%timeit g.transform('mean')

297 µs ± 32.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


# **Transform vs Apply**

* The ***transform*** method is generally ***more efficient than apply*** when we are performing a simple element-wise operation on each group
* The ***apply method*** is ***more flexible but can be slower than transform*** for certain operations.
    * apply method allows more complex operations, and the function passed to apply can return a scalar, a Series, or a DataFrame.
* if the operation is simple, ***transform is often faster***.

# **10.5 Pivot Tables and Cross-Tabulation**
A ***pivot table*** is a ***data summarization*** tool

In [167]:
# From previous tipping dataset
tips

total_bill   tip smoker   day    time  size   tip_pct  tip_pct_times100
0         16.99  1.01     No   Sun  Dinner     2  0.059447          5.944673
1         10.34  1.66     No   Sun  Dinner     3  0.160542         16.054159
2         21.01  3.50     No   Sun  Dinner     3  0.166587         16.658734
3         23.68  3.31     No   Sun  Dinner     2  0.139780         13.978041
4         24.59  3.61     No   Sun  Dinner     4  0.146808         14.680765
..          ...   ...    ...   ...     ...   ...       ...               ...
239       29.03  5.92     No   Sat  Dinner     3  0.203927         20.392697
240       27.18  2.00    Yes   Sat  Dinner     2  0.073584          7.358352
241       22.67  2.00    Yes   Sat  Dinner     2  0.088222          8.822232
242       17.82  1.75     No   Sat  Dinner     2  0.098204          9.820426
243       18.78  3.00     No  Thur  Dinner     2  0.159744         15.974441

[244 rows x 8 columns]

In [169]:
# pivot table - getting mean for each column
# A groupby function
# aggfunc='mean' -> this is by default
tips.pivot_table(index = ['day', 'smoker'], aggfunc='sum')

<ipython-input-169-27eeed0cb83a>:4: FutureWarning: pivot_table dropped a column because it failed to aggregate. This behavior is deprecated and will raise in a future version of pandas. Select only the columns that can be aggregated.
  tips.pivot_table(index = ['day', 'smoker'], aggfunc='sum')


size     tip   tip_pct  tip_pct_times100  total_bill
day  smoker                                                      
Fri  No         9   11.25  0.606602         60.660177       73.68
     Yes       31   40.71  2.621746        262.174578      252.20
Sat  No       115  139.63  7.112145        711.214459      884.78
     Yes      104  120.77  6.212055        621.205474      893.62
Sun  No       167  180.57  9.126438        912.643775     1168.88
     Yes       49   66.82  3.557756        355.775601      458.28
Thur No       112  120.32  7.213414        721.341368      770.09
     Yes       40   51.51  2.785676        278.567563      326.24

In [170]:
# Another example - getting mean for size and tip_pct columns
# Complex but very useful
tips.pivot_table(index = ['time', 'day'], columns='smoker',
                 values=['tip_pct', 'size'], aggfunc='sum')

size          tip_pct          
smoker          No    Yes        No       Yes
time   day                                   
Dinner Fri     6.0   20.0  0.418867  1.488126
       Sat   115.0  104.0  7.112145  6.212055
       Sun   167.0   49.0  9.126438  3.557756
       Thur    2.0    NaN  0.159744       NaN
Lunch  Fri     3.0   11.0  0.187735  1.133620
       Thur  110.0   40.0  7.053669  2.785676

## **Cross-Tabulations: Crosstab**
A ***special pivot table*** to compute ***group frequencies***

In [ ]:
# Import a specific class or object called StringIO from the io module
# allows us to read from or write to strings as if they were files
from io import StringIO

In [ ]:
# data consists of strings
data = """Sample  Nationality  Handedness
1   USA  Right-handed
2   Japan    Left-handed
3   USA  Right-handed
4   Japan    Right-handed
5   Japan    Left-handed
6   Japan    Right-handed
7   USA  Right-handed
8   USA  Left-handed
9   Japan    Right-handed
10  USA  Right-handed"""

In [ ]:
# Create a table
# sep="\s+" -> indicates that one or more whitespace characters are used as the separator between columns
# We called this as raw dataframe
data = pd.read_table(StringIO(data), sep="\s+")
data

Sample Nationality    Handedness
0       1         USA  Right-handed
1       2       Japan   Left-handed
2       3         USA  Right-handed
3       4       Japan  Right-handed
4       5       Japan   Left-handed
5       6       Japan  Right-handed
6       7         USA  Right-handed
7       8         USA   Left-handed
8       9       Japan  Right-handed
9      10         USA  Right-handed

In [ ]:
# Crosstab example 1 - find group frequencies
# summarize by nationality and handedness
# Secondary information by doing some summarization
# margins=True -> to ensure sum of rows and columns are the same
# provide a quick overview of the overall distribution of data across the different combinations of data
pd.crosstab(data['Nationality'], data['Handedness'], margins=True)

Handedness   Left-handed  Right-handed  All
Nationality                                
Japan                  2             3    5
USA                    1             4    5
All                    3             7   10

In [ ]:
# tipping dataset
tips

total_bill   tip smoker   day    time  size   tip_pct  tip_pct_times100
0         16.99  1.01     No   Sun  Dinner     2  0.059447          5.944673
1         10.34  1.66     No   Sun  Dinner     3  0.160542         16.054159
2         21.01  3.50     No   Sun  Dinner     3  0.166587         16.658734
3         23.68  3.31     No   Sun  Dinner     2  0.139780         13.978041
4         24.59  3.61     No   Sun  Dinner     4  0.146808         14.680765
..          ...   ...    ...   ...     ...   ...       ...               ...
239       29.03  5.92     No   Sat  Dinner     3  0.203927         20.392697
240       27.18  2.00    Yes   Sat  Dinner     2  0.073584          7.358352
241       22.67  2.00    Yes   Sat  Dinner     2  0.088222          8.822232
242       17.82  1.75     No   Sat  Dinner     2  0.098204          9.820426
243       18.78  3.00     No  Thur  Dinner     2  0.159744         15.974441

[244 rows x 8 columns]

# ***Crosstab always have two comparisons***:
* list A vs List B
* [tips['time'], tips['day']] as List A
* tips['smoker'] as List B

In [ ]:
# Crosstab example 2
pd.crosstab([tips['time'], tips['day']], tips['smoker'], margins=True)

smoker        No  Yes  All
time   day                
Dinner Fri     3    9   12
       Sat    45   42   87
       Sun    57   19   76
       Thur    1    0    1
Lunch  Fri     1    6    7
       Thur   44   17   61
All          151   93  244

# **That's all for the day!!!**

See you all next week :)